### ISSUE TITLE
Allow JIDs to be used as key in dicts

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2770]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#2770|
|Type|enhancement|
|Created|2007-08-03 11:48:56Z|

This requires the JID object to be immutable and hashable.

<details><summary>Searchable metadata</summary>

```
trac-id__2770 2770
type__enhancement enhancement
reporter__ralphm ralphm
priority__highest highest
milestone__Words_0_6 Words-0.6
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1186141736000000 1186141736000000
changetime__1186502900000000 1186502900000000
version__None None
owner__ 
cc__ralphm cc__therve cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm removed owner|
|-|-|

Worked about the complete module and tests. Also made JID objects immutable and hashable so they can be used in sets and as dict keys.

Please review [branches/jid-2770.](branches/jid-2770.)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @ralphm|
|-|-|

 * the comparison with other objects is not tested (```return NotImplemented``` of ```__eq__``` and ```__ne___```)
 * ```__unicode__``` is not tested
 * pyflakes errors:
```
twisted/words/protocols/jabber/jid.py:14: 'defer' imported but unused
twisted/words/protocols/jabber/jid.py:14: 'protocol' imported but unused
twisted/words/protocols/jabber/jid.py:14: 'reactor' imported but unused
twisted/words/protocols/jabber/jid.py:15: 'domish' imported but unused
twisted/words/protocols/jabber/jid.py:15: 'utility' imported but unused
twisted/words/protocols/jabber/jid.py:17: 'string' imported but unused
twisted/words/test/test_jabberjid.py:56: redefinition of fuction 'test_doubleAt' from line 47
```
 (the second 'test_doubleAt' overrides the first).
 * None of the new methods of JID has docstring
 * the assert in ```__init__``` should be replace by a test and a ```RuntimeError```. Also, str and tuple are pretty terrible names for variables, but I guess it can't be changed transparently :/

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm set owner to @therve|
|-|-|

Addressed review comments.

I could maybe do away with `str` as an argument name to `__init__`, but I am not sure about how much `tuple` is being used. I agree their naming is unfortunate.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @ralphm|
|-|-|

OK. There are some blank lines in the docstrings of the JID methods that should be removed. Also, it would be cool to limit the file to 80 columns.

I think the immutability is blocking this branch for now. I'll give my go if you remove it, but then it may need another ticket to give a verdict on that.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_7"></a>|@ralphm set owner to @therve|
|-|-|

I removed the immutability, fixed the docstrings and created a new ticket for it: [#2776](https://github.com/twisted/twisted/issues/2776).

Please review again.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @ralphm|
|-|-|

Fine please merge.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_9"></a>|@ralphm set status to `closed`|
|-|-|

(In [20987]) Make JIDs hashable for using them as dict keys and clean up of the module.

Author: ralphm
Reviewer: therve
Fixes [#2770](https://github.com/twisted/twisted/issues/2770).

 * Adds `__unicode__` for getting a string representation of a JID in a
more pythonic way.
 * Cleans up the comparison operator implementation with
regard to comparing against non-JID objects.
 * Better docstrings.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
